src/libopensc/card-fineid.c

```diff
@@ -0,0 +1,997 @@
+		if(orglen > ilen) {
+			rv = fineid_change_security_env(card);
+			LOG_TEST_RET(card->ctx, rv, "Security env change with new algorithm failed");
+		}
+	}
+
+	/* If data does not fit in block in length */
+	if(ilen>blklen) {
+		for (ii=0; ii<ilen-blklen; ii+=blklen) {
+			/* Security operation hash
+			 *   INS 2A sec op
+			 *   P1  90 hash
+			 *   P2  80 one block of data
+			 */
+			sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0x2A, 0x90, 0x80);
+			apdu.datalen = blklen;
+			apdu.data = instr+ii;
+			apdu.lc = blklen;
+
+			sc_log(card->ctx, "Iterating at offset %lu", ii);
+			rv = sc_transmit_apdu(card, &apdu);
+			LOG_TEST_RET(card->ctx, rv, "APDU transmit failed");
+			rv = sc_check_sw(card, apdu.sw1, apdu.sw2);
+			LOG_TEST_RET(card->ctx, rv, "Block send with offset failed");
+		}
+	}
+
+	reqlen = ilen-ii+2;
+
+	memcpy(&req[2], instr+ii, ilen-ii);
+
+	/* Decide who will/has calculate(d) the hash */
+	if(_driver_data->algorithm_flags & SC_ALGORITHM_RSA_HASH_MD5_SHA1)
+	    req[0] = FINEID_HASHING_EXTERNALLY;
+	else if(fineid_get_algo(_driver_data->algorithm_flags) == FINEID_ALGO_HIGH_NA)
+		req[0] = FINEID_HASHING_BY_CARD;
+	else
+		req[0] = FINEID_HASHING_EXTERNALLY;
+
+	req[1] = ilen-ii;
+
+	sc_log(card->ctx, "Finalizing at offset %lu", ii);
+
+	/* Security operation hash
+	 *   INS 2A sec op
+	 *   P1  90 hash
+	 *   P2  A0 last data block
+	 */
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0x2A, 0x90, 0xA0);
+	apdu.datalen = reqlen;
+	apdu.data = req;
+	apdu.lc = reqlen;
+
+	rv = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(card->ctx, rv, "APDU transmit failed");
+	rv = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	LOG_TEST_RET(card->ctx, rv, "Last block send failed");
+
+	/* Security operation sign
+	 *   INS 2A sec op
+	 *   P1  9E sign, signature is returned in response
+	 *   P2  9A
+	 */
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_2_SHORT, 0x2A, 0x9E, 0x9A);
+	apdu.le = olen > 256 ? 256 : olen;
+	apdu.resp = resp;
+	apdu.resplen = olen;
+
+	rv = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(card->ctx, rv, "APDU transmit failed");
+	rv = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	LOG_TEST_RET(card->ctx, rv, "Signature receiving failed");
+
+	if (apdu.resplen > olen) {
+		sc_log(card->ctx,
+		       "Compute signature failed: invalid response length %"SC_FORMAT_LEN_SIZE_T"u",
+		       apdu.resplen);
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_CARD_CMD_FAILED);
+	}
+
+	memcpy(out, apdu.resp, apdu.resplen);
+
+	LOG_FUNC_RETURN(card->ctx, apdu.resplen);
+}
+
+
+/* Return the default AAK for this type of card */
+static int
+fineid_get_default_key(struct sc_card *card, struct sc_cardctl_default_key *data)
+{
+	LOG_FUNC_RETURN(card->ctx, SC_ERROR_NO_DEFAULT_KEY);
+}
+
+
+static int
+fineid_card_ctl(struct sc_card *card, unsigned long cmd, void *ptr)
+{
+	switch (cmd) {
+	case SC_CARDCTL_GET_DEFAULT_KEY:
+		return fineid_get_default_key(card,
+				(struct sc_cardctl_default_key *) ptr);
```

Comment:
did you test pkcs15-init with your card? If not, don't implement SC_CARDCTL_GET_DEFAULT_KEY